h/api/models.py

```diff
@@ -166,6 +175,113 @@ def title(self):
+          the scheme (http(s)://) and www parts, e.g. "example.com".
+          If it's a local file:// uri then the filename is used as the domain.
+          If the domain is too long it is truncated with &hellip;.
+
+        The ({domain}) part will be missing if it wouldn't be any different
+        from the {link_text} part.
+
+        The href="{href}" will be missing if there's no http(s) uri to link to
+        for this annotation's document.
+
+        User-supplied values are escaped so the string is safe for raw
+        rendering (the returned string is actually a jinja2.Markup object and
+        won't be escaped by Jinja2 when rendering).
+
+        """
+        uri = jinja2.escape(self.uri) or ""
+
+        if (uri.lower().startswith("http://") or
+                uri.lower().startswith("https://")):
+            href = uri
+        else:
+            href = ""
+
+        if self.title:
+            title = jinja2.escape(self.title)
+            link_text = title
+            if is_file(uri):
+                domain = jinja2.escape(self.filename)
+            else:
+                domain = urlparse.urlparse(uri).hostname
+        else:
+            if is_file(uri):
+                title = urllib2.unquote(jinja2.escape(self.filename))
+                link_text = title
+                domain = ""
+            else:
+                title = urllib2.unquote(uri)
+                parts = urlparse.urlparse(uri)
+                link_text = urllib2.unquote(parts.netloc + parts.path)
+                domain = ""
+        if domain == title:
+            domain = ""
+
+        def truncate(content, length=50):
+            """Truncate the given string to at most length chars."""
+            if len(content) <= length:
+                return content
+            else:
+                return content[:length] + jinja2.Markup("&hellip;")
+
+        if link_text:
+            link_text = truncate(link_text)
+
+        if domain:
+            domain = truncate(domain)
+
+        assert title, "The title should never be empty"
+        assert link_text, "The link text should never be empty"
+        if href and domain:
+            link = ('<a href="{href}" title="{title}">{link_text}</a> '
+                    '({domain})'.format(href=href, title=title,
+                                        link_text=link_text, domain=domain))
+        elif domain and not href:
+            link = ('<a title="{title}">{link_text}</a> ({domain})'.format(
+                title=title, link_text=link_text, domain=domain))
+        elif href and not domain:
+            link = '<a href="{href}" title="{title}">{link_text}</a>'.format(
+                href=href, title=title, link_text=link_text)
+        elif (not href) and (not domain):
+            link = '<a title="{title}">{link_text}</a>'.format(
+                title=title, link_text=link_text)
+        else:
+            assert False, "We should never get here"
```

Comment:
This assertion is unnecessary if you just reduce the `if` clause to its minimal form.

```python
if href and domain:
    ...
elif domain: # domain and no href
    ...
elif href: # href and no domain
    ...
else: # no domain and no href
    ...
```